File: src/secsentry/scan/aggregate.py
"""Collapse occurrences into unique secrets.

One fingerprint is one secret, however many files and commits contain it.
A report that lists 500 identical lines is not a report.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime, timezone

from secsentry.models import Finding
from secsentry.rotation import rotation_hint

SEVERITY_RANK = {"low": 1, "medium": 2, "high": 3, "critical": 4}


def rank(severity: str) -> int:
    return SEVERITY_RANK.get(severity.lower(), 0)


def _parse(ts: str) -> datetime | None:
    if not ts:
        return None
    try:
        return datetime.fromisoformat(ts)
    except ValueError:
        return None


@dataclass
class Secret:
    """One unique credential and everywhere it appears."""

    fingerprint: str
    secret_type: str
    severity: str
    masked: str
    confidence: float
    occurrences: list[Finding] = field(default_factory=list)
    first_seen: str = ""
    last_seen: str = ""
    introduced_by: str = ""
    introduced_email: str = ""
    introduced_commit: str = ""
    still_in_head: bool | None = None
    age_days: int | None = None
    rotation: str = ""
    why: list[str] = field(default_factory=list)

    @property
    def paths(self) -> list[str]:
        seen: list[str] = []
        for o in self.occurrences:
            if o.path not in seen:
                seen.append(o.path)
        return seen

    def to_dict(self) -> dict:
        return {
            "fingerprint": self.fingerprint,
            "secret_type": self.secret_type,
            "severity": self.severity,
            "masked": self.masked,
            "confidence": round(self.confidence, 3),
            "still_in_head": self.still_in_head,
            "first_seen": self.first_seen,
            "last_seen": self.last_seen,
            "age_days": self.age_days,
            "introduced_by": self.introduced_by,
            "introduced_email": self.introduced_email,
            "introduced_commit": self.introduced_commit,
            "rotation": self.rotation,
            "why": self.why,
            "occurrence_count": len(self.occurrences),
            "paths": self.paths,
            "occurrences": [o.to_dict() for o in self.occurrences],
        }


def group(findings: list[Finding]) -> list[Secret]:
    by_fp: dict[str, Secret] = {}

    for f in findings:
        s = by_fp.get(f.fingerprint)
        if s is None:
            s = Secret(
                fingerprint=f.fingerprint,
                secret_type=f.secret_type,
                severity=f.severity,
                masked=f.masked,
                confidence=f.confidence,
                why=list(f.why),
                rotation=rotation_hint(f.secret_type),
            )
            by_fp[f.fingerprint] = s
        s.occurrences.append(f)
        if rank(f.severity) > rank(s.severity):
            s.severity = f.severity
        s.confidence = max(s.confidence, f.confidence)
        if f.still_in_head:
            s.still_in_head = True
        elif s.still_in_head is None and f.still_in_head is False:
            s.still_in_head = False

    now = datetime.now(timezone.utc)
    for s in by_fp.values():
        dated = [(o, _parse(o.timestamp)) for o in s.occurrences]
        dated = [(o, d) for o, d in dated if d is not None]
        if dated:
            dated.sort(key=lambda pair: pair[1])
            earliest, first_dt = dated[0]
            latest, last_dt = dated[-1]
            s.first_seen = earliest.timestamp
            s.last_seen = latest.timestamp
            s.introduced_by = earliest.author
            s.introduced_email = earliest.author_email
            s.introduced_commit = earliest.commit
            s.age_days = max((now - first_dt).days, 0)

    return sorted(
        by_fp.values(),
        key=lambda s: (-rank(s.severity), -s.confidence, s.secret_type),
    )

File: internal/scan/aggregate.go
package scan

import (
	"sort"
	"time"

	"github.com/umeraamir69/secsentry/internal/model"
	"github.com/umeraamir69/secsentry/internal/rotate"
)

type Secret struct {
	Fingerprint      string          `json:"fingerprint"`
	SecretType       string          `json:"secret_type"`
	Severity         string          `json:"severity"`
	Masked           string          `json:"masked"`
	Confidence       float64         `json:"confidence"`
	Occurrences      []model.Finding `json:"occurrences"`
	FirstSeen        string          `json:"first_seen"`
	LastSeen         string          `json:"last_seen"`
	IntroducedBy     string          `json:"introduced_by"`
	IntroducedEmail  string          `json:"introduced_email"`
	IntroducedCommit string          `json:"introduced_commit"`
	StillInHead      *bool           `json:"still_in_head"`
	AgeDays          *int            `json:"age_days"`
	Rotation         string          `json:"rotation"`
	Why              []string        `json:"why"`
}

func (s Secret) Paths() []string {
	seen := map[string]bool{}
	var out []string
	for _, o := range s.Occurrences {
		if !seen[o.Path] {
			seen[o.Path] = true
			out = append(out, o.Path)
		}
	}
	return out
}

func (s Secret) OccurrenceCount() int { return len(s.Occurrences) }

func parseTS(ts string) (time.Time, bool) {
	if ts == "" {
		return time.Time{}, false
	}
	for _, layout := range []string{time.RFC3339, time.RFC3339Nano, "2006-01-02T15:04:05"} {
		if t, err := time.Parse(layout, ts); err == nil {
			return t, true
		}
	}
	return time.Time{}, false
}

func Group(findings []model.Finding) []Secret {
	byFP := map[string]*Secret{}
	order := []string{}
	for _, f := range findings {
		s, ok := byFP[f.Fingerprint]
		if !ok {
			s = &Secret{
				Fingerprint: f.Fingerprint,
				SecretType:  f.SecretType,
				Severity:    f.Severity,
				Masked:      f.Masked,
				Confidence:  f.Confidence,
				Why:         append([]string{}, f.Why...),
				Rotation:    rotate.Hint(f.SecretType),
			}
			byFP[f.Fingerprint] = s
			order = append(order, f.Fingerprint)
		}
		s.Occurrences = append(s.Occurrences, f)
		if model.Rank(f.Severity) > model.Rank(s.Severity) {
			s.Severity = f.Severity
		}
		if f.Confidence > s.Confidence {
			s.Confidence = f.Confidence
		}
		if f.StillInHead != nil && *f.StillInHead {
			t := true
			s.StillInHead = &t
		} else if s.StillInHead == nil && f.StillInHead != nil && !*f.StillInHead {
			fval := false
			s.StillInHead = &fval
		}
	}

	now := time.Now().UTC()
	out := make([]Secret, 0, len(order))
	for _, fp := range order {
		s := byFP[fp]
		type pair struct {
			f model.Finding
			t time.Time
		}
		var dated []pair
		for _, o := range s.Occurrences {
			if t, ok := parseTS(o.Timestamp); ok {
				dated = append(dated, pair{o, t})
			}
		}
		if len(dated) > 0 {
			sort.Slice(dated, func(i, j int) bool { return dated[i].t.Before(dated[j].t) })
			earliest, latest := dated[0], dated[len(dated)-1]
			s.FirstSeen = earliest.f.Timestamp
			s.LastSeen = latest.f.Timestamp
			s.IntroducedBy = earliest.f.Author
			s.IntroducedEmail = earliest.f.AuthorEmail
			s.IntroducedCommit = earliest.f.Commit
			age := int(now.Sub(earliest.t).Hours() / 24)
			if age < 0 {
				age = 0
			}
			s.AgeDays = &age
		}
		out = append(out, *s)
	}
	sort.Slice(out, func(i, j int) bool {
		ri, rj := model.Rank(out[i].Severity), model.Rank(out[j].Severity)
		if ri != rj {
			return ri > rj
		}
		if out[i].Confidence != out[j].Confidence {
			return out[i].Confidence > out[j].Confidence
		}
		return out[i].SecretType < out[j].SecretType
	})
	return out
}
